### ISSUE TITLE
Calling Cast API functions from C# .NET Framework

### ISSUE THREAD

**Issue by RobKondratowicz** [clariusdev/cast#90]
I am interested in using the Cast API in our C# WPF .NET Framework application. I was able to compile and call some of these functions however they do not appear to be working properly. We currently do not have a probe to work with at the moment, however I wanted to get a head start on integrating with the API. We will be getting a probe sometime this month.
I am on Windows 10 using Visual Studio 2019, using version 10.0.0 of the cast.dll.

I am using P/Invoke with the DLLImport attribute to import the functions as follows
```csharp
[DllImport("cast.dll", CallingConvention = CallingConvention.Cdecl)]
public static extern int cusCastInit(
    int argc,
    string[] argv,
    [MarshalAs(UnmanagedType.LPStr)] string dir,
    CusNewProcessedImageFn newProcessedImage,
    CusNewRawImageFn newRawImage,
    CusNewSpectralImageFn newSpectralImage,
    CusFreezeFn freeze,
    CusButtonFn btn,
    CusProgressFn progress,
    CusErrorFn err,
    int width,
    int height
);

[DllImport("cast.dll", CallingConvention = CallingConvention.Cdecl)]
public static extern int cusCastConnect([MarshalAs(UnmanagedType.LPStr)] string ipAddress, uint port, [MarshalAs(UnmanagedType.LPStr)] string cert, CusReturnFn fn);
```

And the functions are getting called as follows
```csharp
public static void Main(String[] args)
{
    var ret = 10000;
    ret = CusCast.cusCastInit(1, new string[]{"ClariusMain.exe"}, Environment.CurrentDirectory, NewProcessedImage, NewRawImage, NewSpectralImage, Freeze, Btn, Progress, Err, 640, 480);
    Console.WriteLine($"ClariusMain: cusCastInit returns : {ret} ");

    int UDPPort = 10000;
    int ret2 = 10000;
    ret2 = CusCast.cusCastConnect("localhost", 5679, "research", portNum => { UDPPort = portNum; });
    Console.WriteLine($"ClariusMain: cusCastConnect returns: {ret2} and the UDP port is {UDPPort}");
}
```

The output from the code above prints out the following
```
QWindowsContext: OleInitialize() failed:  "COM error 0xffffffff80010106 RPC_E_CHANGED_MODE (Cannot change thread mode after it is set.)"
ClariusMain: cusCastInit returns : 0
ClariusMain: cusCastConnect returns: 0 and the UDP port is 10000
Press any key to continue . . .
```

If i'm not mistaken the `QWindowsContext: OleInitialize() failed:` error is ok and the rest of the API should still function as normal.
the other more concerning issue is that the `cusCastConnect` function appears to be getting called properly and is returning 0 (Success) even though it should fail since we do not have a probe on the network.

If I try calling `cusCastConnect` without first calling `cusCastInit` then `cusCastConnect` will return -1 which is expected. So I believe that I am importing these functions into C# correctly. I have also tried calling `cusCastConnect` with different IP address, both valid and invalid, as well with different ports. each way I get a similar result.

Do you know why the `cusCastConnect` function is reporting that the connection was successful even without a probe available?

P.S. Is there any way to use one of the virtual probes with the Cast API for testing?

**Comment by clariusk** [comment]
yes, this makes sense, from the documentation (sorry if it is ambiguous)

```
/// retval 0 the connection attempt was successful
/// retval -1 the connection attempt was not successful
```

meaning, the call was successful (which relies on init() succeeding), but since this is an async operation, you are required to get the network connection results through the callback function provided.

A virtual probe would be nice, we can think of deploying one in the future, thanks!

**Comment by RobKondratowicz** [comment]
Ah ok, That is making sense. I was able to make the simple program wait for the the callback to be run, and sure enough it was called after a little bit of time passed and it set the UDP port to 0. I also decided to output the Error message from the error message callback, and that is also reporting that the connection failed in the expected manor.

Just for clarity sake. The return value of `cusCastConnect` is 0 if the function it self was successfully called and -1 if it failed.
Then the callback that is supplied is run with a valid port number if the connection was successful and 0 if it was unsuccessful.
Is this correct?
